ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTConverter.java

```diff
@@ -756,17 +737,8 @@ public ASTNode visitInputRef(RexInputRef inputRef) {
     @Override
     public ASTNode visitLiteral(RexLiteral literal) {
 
-      if (RexUtil.isNull(literal) && literal.getType().getSqlTypeName() != SqlTypeName.NULL
-          && rexBuilder != null) {
-        // It is NULL value with different type, we need to introduce a CAST
-        // to keep it
-        if(nullLiteralMap.containsKey(literal)) {
-          return ASTBuilder.literal(literal);
-        }
-        nullLiteralMap.put(literal, true);
-        RexNode r = rexBuilder.makeAbstractCast(literal.getType(), literal);
-
-        return r.accept(this);
+      if (RexUtil.isNull(literal) && literal.getType().getSqlTypeName() != SqlTypeName.NULL) {
+        return createNullField(literal.getType());
       }
       return ASTBuilder.literal(literal);
```

Comment:
Maybe in a follow up we could add some unit tests for this visitor and in particular for this method so that we can test all supported null/not null types. Not strictly needed as part of this change.